### DISCUSSION TITLE
grid, layout, columns, templates

### DISCUSSION THREAD

**Discussion by brianeckmann** [microsoft/atlas-design#74] [Q&A]
I'm trying to understand how we're calculating the content widths and margin left/right at different breakpoints so I can help map templates into Figma for responsive design. The documentation is showing gap values in pixels/rem, but on the site I'm seeing percentages and a mix of classes I can't quite follow. It's not a 1:1 to get things into Figma but we can get close enough to translate the redlines.

I also need to set guidelines on what comps dev needs from design. I.e., do we need to deliver one at desktop, tablet, and mobile? Do we need widescreen? Or do our components solve all of this for us?

This is all in the name of creating Figma templates for designers with a grid built-in. But I need to know for different screen widths how the grid is calculated. What I have so far is the gap value (1.5 rem), so then do I set a max-width value at each breakpoint? From there Figma can split it into 12 columns.

Hopefully I'm making sense :D

**Comment by wibjorn** [comment]
Since we seem to be generally discussing layout methods, here are some layout and spacing related concepts to riff on what you've put above.

**Site layout** -> this is how the major players on the page fit together. We can and should, in my mind, create a new implementation for this based on cssgrid. This deals with how the major elements (header, footer, table of contents, sidebar, etc) on a page resize and how much space they take up.

**Columns** -> This is a standard old school styles 12 column grid. The elements that it contains can vary, but they're widths are in percentages. This isn't the most modern way of doing layout, but it works for many cases. _This does not set max widths, 12 columns just takes up 100% of the container._ As we've talked about in the past, gaps between columns can be 0, .75rem, and 1.5rem.

**Container** -> This constrains widths of things, particularly of wide containers. _This can set max widths._

**Page gutters** -> This is what the old .uhf-container class on docs did. It made things line up with the side of the page. I am not entirely sure we need this, but as a concept it's definitely worth discussing.

**CSS Grid** -> This is a [modern series of css rules](https://css-tricks.com/snippets/css/complete-guide-grid/) that provide us with a ton of flexibility. We do have some limitations until IE is no longer supported, but we can do most of what's here. On Docs and Learn, the most common case of it is on the [content browsers](https://docs.microsoft.com/en-us/learn/browse/). This is super handy because elements within the grid can automatically resize themselves given a set of minimums and maximum widths. This is chief among a host of reasons with css is great.

CSS Feature -> Clamp! @5t3ph. With [clamp](https://developer.mozilla.org/en-US/docs/Web/CSS/clamp()) You can set a minimum, preferred and maximum values for an elements width. This is flexible, modern and simple.

Regarding this:

> I also need to set guidelines on what comps dev needs from design. I.e., do we need to deliver one at desktop, tablet, and mobile? Do we need widescreen? Or do our components solve all of this for us?

For a component design, I would not think it would be necessary to provide designs for different widths unless that component has different appearances at different widths. For anything where page reflow occurs and is a crucial part of the design (as in a page-layout specific design, perhaps) I would expect there to be designs for different.

**Reply by 5t3ph** [reply]
Container queries are coming officially as well but even without them we can get super close with flex/grid for what you usually want to do for basic flips of orientation (ex media next to content in large container, media on top in small container). By which I mean - components could start to consider their _ideal_ space/behavior vs. "at this breakpoint I do this". So we could have page layouts that have some breakpoint behavior like to be more firm about number of card columns for examle, but components could respond to _container width_.

I understand your ask is to help you design Figma templates - so perhaps those focus on wide view to more indicate general page layout, since the children should be typically using "approved" components which could have their layout behavior built in. This is a conversation I'm super eager to have because it's the area of CSS I nerd out about the _most_ haha

**Reply by brianeckmann** [reply]
I'm all for the latest and greatest, sounds like Stephanie and Will are already onto a viable solution.

I don't want to over-complicate anything, a 12 column grid works for design. Would the CSS grid still operate with 12-columns? I know we can pretty much do whatever we want with it, but we still need some kind of consistent layouts to use or the page designs will get crazy. Any additional benefits we inherit to be able to build custom layouts quickly will be exciting to know about too.

We have the breakpoints identified but I think we'll need some more values like max-content widths and left/right margins to create Figma templates. At our breakpoints we'll also need to determine how our layouts adjust, example being on tablet if we go from 12 to 6 columns.

**Reply by brianeckmann** [reply]
![image](https://user-images.githubusercontent.com/82234828/114721653-24a42080-9d07-11eb-93fb-f7231b4b4513.png)

FYI this is the kind of varying alignment I'd like to clean up and get in sync with ACOM on to a similar layout

**Reply by wibjorn** [reply]
@brianeckmann  That makes sense to me, although I think that we don't need to spend too much time on that early. To me, it makes more sense to focus on components and keep alignment and containers very simple. With that said, is the eventual goal for 
"containers" that all major sections should align vertically with the header?

**Reply by brianeckmann** [reply]
Replying to this 3 weeks later...yes that's the case. The nav/header should align to the content within the same container. Everything should align to the left and right edges--and it's okay to have something centered within those boundaries.

I would think this is one of the first things to get sorted out if we're going to bring over classes such .content, .nav-bar, .container etc.

**Reply by wibjorn** [reply]
@brianeckmann @courtny Sure. A few clarifications.

- .container (now) sets a max width
- .nav-bar has a different one set.
- the .content container (which contains article text, for example) really shouldn't match the nav bar, because it will be very, very wide.

That said, I think we'll have to decide the following:

1) Do we want to change the site's gutter values from what they are on the Docs site now?
  - mobile is currently 16px to either side.
  - tablet is currently 24px to either side.
  - desktop is 5vw (5 percent of viewport width, not factoring in scrollbars) to either side.
  - above that a max width is set, and gutters are just using margin: auto, which centers content.
  - seems like we could improve upon this, but knowing how the site's layout might evolve will help inform us here.
2) Do we want to impose/change a maximum width (1600px) is what's done now?
3) Are there major site layout changes coming? If so, should we be forward thinking in our values?
